spec/abstract/Schedule_A.thy

```diff
@@ -66,36 +63,47 @@ definition
      modify (\<lambda>s. s \<lparr> cur_thread := thread \<rparr>)
    od"
 
-class state_ext_sched = state_ext +
-  fixes schedule :: "(unit,'a) s_monad"
+definition
+  next_domain :: "(unit, 'z::state_ext) s_monad" where
+  "next_domain \<equiv> do
+    modify (\<lambda>s.
+      let domain_index' = (domain_index s + 1) mod length (domain_list s) in
+      let next_dom = (domain_list s)!domain_index'
+      in s\<lparr> domain_index := domain_index',
+            cur_domain := fst next_dom,
+            domain_time := snd next_dom\<rparr>);
+    do_extended_op $ modify (\<lambda>s. s \<lparr>work_units_completed := 0\<rparr>)
+  od"
 
-definition choose_thread :: "det_ext state \<Rightarrow> (unit \<times> det_ext state) set \<times> bool" where
-"choose_thread \<equiv>
-      do
-        d \<leftarrow> gets cur_domain;
-        queues \<leftarrow> gets (\<lambda>s. ready_queues s d);
-        if (\<forall>prio. queues prio = []) then (switch_to_idle_thread)
-        else (guarded_switch_to (hd (max_non_empty_queue queues)))
-      od"
+definition
+  dec_domain_time :: "(unit, 'z::state_ext) s_monad" where
+  "dec_domain_time = modify (\<lambda>s. s\<lparr>domain_time := domain_time s - 1\<rparr>)"
+
+definition max_non_empty_queue :: "(priority \<Rightarrow> ready_queue) \<Rightarrow> ready_queue" where
+  "max_non_empty_queue queues \<equiv> queues (Max {prio. queues prio \<noteq> []})"
+
+definition choose_thread :: "(unit, 'z::state_ext) s_monad" where
+  "choose_thread \<equiv> do
+     d \<leftarrow> gets cur_domain;
+     queues \<leftarrow> gets (\<lambda>s. ready_queues s d);
+     if \<forall>prio. queues prio = []
+     then switch_to_idle_thread
+     else guarded_switch_to (hd (max_non_empty_queue queues))
+   od"
 
 text \<open>
   Determine whether given priority is highest among queued ready threads in given domain.
   Trivially true if no threads are ready.\<close>
 definition
-  is_highest_prio :: "domain \<Rightarrow> priority \<Rightarrow> det_ext state \<Rightarrow> bool"
+  is_highest_prio :: "domain \<Rightarrow> priority \<Rightarrow> 'z::state_ext state \<Rightarrow> bool"
```

Comment:
that's unexpected, any reason for this change?

Reply:
Now that the scheduler state is part of the base state instead of the extended state, functions like this that previously operated on `det_ext state` can instead be generic. If I didn't make this change then other functions that call `is_highest_prio` would also need to be `det_ext state`, which would unnecessarily force us to still have a deterministic and non-deterministic scheduler.